File: backend/app/models/trade_idea.py
from sqlalchemy import Column, Integer, String, Text, DateTime, Numeric, ForeignKey, Index, func
from sqlalchemy.orm import relationship
from app.models.base import Base


class TradeIdea(Base):
    __tablename__ = 'trade_ideas'

    id = Column(Integer, primary_key=True, index=True)
    symbol = Column(String(20), nullable=False)
    direction = Column(String(10), nullable=False)  # LONG or SHORT
    entry_price_target = Column(Numeric(precision=18, scale=8))
    stop_price = Column(Numeric(precision=18, scale=8))
    target_price = Column(Numeric(precision=18, scale=8))
    thesis = Column(Text)
    timeframe = Column(String(50))
    confidence = Column(String(20))  # LOW / MEDIUM / HIGH
    tags = Column(String(200))
    revisit_date = Column(DateTime)

    # Status machine: draft → active → traded / archived
    status = Column(String(20), nullable=False, default='draft')
    traded_trade_id = Column(Integer, ForeignKey('trades.id'), nullable=True)
    triggered_at = Column(DateTime)

    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(DateTime, nullable=False, server_default=func.now(), onupdate=func.now())

    # Optional FK to Trade (populated when idea is converted to trade)
    traded_trade = relationship('Trade', back_populates='source_idea')

    __table_args__ = (
        Index('ix_trade_ideas_status', 'status'),
        Index('ix_trade_ideas_symbol', 'symbol'),
    )

File: frontend/src/types/tradeIdea.ts
// Trade Idea types — mirror of backend Pydantic schemas
// Backend serializes Decimal prices to strings

export type TradeIdeaStatus = 'draft' | 'active' | 'traded' | 'archived'
export type TradeIdeaDirection = 'LONG' | 'SHORT'
export type TradeIdeaConfidence = 'LOW' | 'MEDIUM' | 'HIGH'

export interface TradeIdeaItem {
  id: number
  symbol: string
  direction: TradeIdeaDirection
  entry_price_target: string | null
  stop_price: string | null
  target_price: string | null
  thesis: string | null
  timeframe: string | null
  confidence: TradeIdeaConfidence | null
  tags: string | null
  revisit_date: string | null
  status: TradeIdeaStatus
  traded_trade_id: number | null
  triggered_at: string | null
  created_at: string
  updated_at: string
}

export interface TradeIdeaListResponse {
  total: number
  items: TradeIdeaItem[]
}

export interface TradeIdeaCreatePayload {
  symbol: string
  direction: TradeIdeaDirection
  entry_price_target?: string | null
  stop_price?: string | null
  target_price?: string | null
  thesis?: string | null
  timeframe?: string | null
  confidence?: TradeIdeaConfidence | null
  tags?: string | null
  revisit_date?: string | null
  status?: TradeIdeaStatus
}

export interface TradeIdeaUpdatePayload {
  symbol?: string | null
  direction?: TradeIdeaDirection | null
  entry_price_target?: string | null
  stop_price?: string | null
  target_price?: string | null
  thesis?: string | null
  timeframe?: string | null
  confidence?: TradeIdeaConfidence | null
  tags?: string | null
  revisit_date?: string | null
  status?: TradeIdeaStatus | null
}

export interface ConvertToTradePayload {
  entry_price?: string | null
  exit_price?: string | null
  quantity?: string | null
  fees?: string | null
  notes?: string | null
}

export interface ConvertToTradeResponse {
  idea: TradeIdeaItem
  trade_id: number | null
}
